### DISCUSSION TITLE
Adventure map, Tree of Knowledge, Right click info window, Display the cost of the tree of knowledge bonus when a recruited hero has already visited the tree of knowledge while under the command of another player

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#9473] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Display the cost of the tree of knowledge bonus in the right click info window when a recruited hero has already visited the tree of knowledge while under the command of another player.

From @zenseii:
> > > If the tree of knowledge has already been visited by the hero, the cost to be paid should also be displayed.
> > 
> > The way I've come to see this mechanic of knowing what bonus objects a hero has visited and used, is that this is a gameplay feature. It answer the player's simple question of "Will there be any use for my hero to visit this location?"
> 
> As this PR is closed, I will open another issue for this.

_Originally posted by @LeHerosInconnu in https://github.com/ihhub/fheroes2/issues/9077#issuecomment-2326150312_

I don't understand the choice of not displaying the cost to be paid in the right click information window.
If the hero has already benefited from the tree of knowledge bonus, the cost to be paid is known, so why not display it?

The player should be able to benefit from this information for his other heroes, this is not just for an individual hero.

What's the point of forcing the user to send another hero to find out the cost to be paid?
It just seems like an annoying micromanagement imposed on the player, who will even consider it a bug because it doesn't work in the same way as it does with his natively recruited heroes when they visit a tree of knowledge.

### Describe the possible solution

If a hero, even one previously under another player's command, has benefited from the tree of knowledge bonus (or has simply visited the tree of knowledge), the cost of accessing the bonus should be indicated in the right click information window.

### Additional info

Related PR: https://github.com/ihhub/fheroes2/pull/9077.
Related discussion: https://github.com/ihhub/fheroes2/discussions/9081.

**Comment by zenseii** [comment]
@LeHerosInconnu, I personally believe that this issue should be converted to a discussion because it is divisive and a big step in changing the original gameplay (it is not simply a QoL thing), as we found out in the previous PR's discussion. At the time of the PR, neither me nor @oleg-derevenetz thought that what this issue suggests should be implemented.

@ihhub, what do you think?

**Comment by ihhub** [comment]
Hi @zenseii , we don't need to make the game too casual for players. If another player visited the tree we shouldn't display extra information about the object for **other** players. I haven't seen such game logic in other games in HOMM series.

**Comment by LeHerosInconnu** [comment]
Hello @zenseii and @ihhub,

> @LeHerosInconnu, I personally believe that this issue should be converted to a discussion because it is divisive and a big step in changing the original gameplay (it is not simply a QoL thing), as we found out in the previous PR's discussion. At the time of the PR, neither me nor @oleg-derevenetz thought that what this issue suggests should be implemented.
> 
> @ihhub, what do you think?

In the original game, additional location information was not displayed; fheroes2 has improved this.
In the original game, basic location information is displayed when an opposing hero is recruited.
For me, it's clear that an opposing hero recruited by the player brings with him the basic information of the locations (displayed in the original game) and complementary information (added in fheroes2) he has already visited during the game.
This in no way changes the gameplay of the original game.

> Hi @zenseii , we don't need to make the game too casual for players. If another player visited the tree we shouldn't display extra information about the object for **other** players. I haven't seen such game logic in other games in HOMM series.

This doesn't make the game any more casual.
The logic of the original game is preserved; the basic location information known by the opposing hero is retained when the player recruits him.
Now that additional location information is displayed with fheroes2, this additional location information known by the recruited opposing hero must also be retained and therefore displayed.

Related discussion: https://github.com/ihhub/fheroes2/discussions/9081.

**Comment by Branikolog** [comment]
Hi!
My 50 cents...
I believe, that all information regarding trees, dwellings, witch huts, shrines refers to a kingdom, rather than a particular hero hired from the tavern. I'm 300% sure all players are going to be confused about the information they couldn't discovered personally.

I'm personally agreed to get such info from allied players, but not by hiring a hero previously owned by an enemy...

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog,

> Hi! My 50 cents... I believe, that all information regarding trees, dwellings, witch huts, shrines refers to a kingdom, rather than a particular hero hired from the tavern. I'm 300% sure all players are going to be confused about the information they couldn't discovered personally.
> 
> I'm personally agreed to get such info from allied players, but not by hiring a hero previously owned by an enemy...

You mention player confusion.
For me, that is confusion:

In fheroes2:

https://github.com/user-attachments/assets/c265afd4-fdb7-4438-a181-9f4916cf0c5b

https://github.com/user-attachments/assets/37eca4e5-ba20-408f-8ffe-1c1ccf1e3f7b

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu 
I love how magic garden works. This is a unique object, that has two states (full and empty) and I really enjoy finding out by myself, that enemy has visited it before, by only looking at its sprite. "Visited"/"not visited" state works here same to every other object in the game - I cannot know, if enemy has visited it. Same, like you can find out that someone ate your ice cream from the box without opening it, by simple looking on some fingerprints on the box. It is just funny and unique quirk of this particular object. =)
Speaking about the tree. I personally rather accept not to show visited status for hero at all, that visited it under the command of the other player, than show the precise information for an object, player never visited before. I think this is more fair. But I expect players would be slightly annoyed by that fact, so the current implementation is OK for me. 
Just my personal opinion. 🙂

**Comment by zenseii** [comment]
I'm not saying we should do this, but it seems that the most logical solution to @LeHerosInconnu's issue is to stop showing that a newly recruited hero has visited a tree under the command of another player, thus forcing the player to visit the tree with that new hero to find this out for themself. However, I believe this would be contrary to how all later heroes games work.

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog and @zenseii,

> Hi, @LeHerosInconnu I love how magic garden works. This is a unique object, that has two states (full and empty) and I really enjoy finding out by myself, that enemy has visited it before, by only looking at its sprite. "Visited"/"not visited" state works here same to every other object in the game - I cannot know, if enemy has visited it. Same, like you can find out that someone ate your ice cream from the box without opening it, by simple looking on some fingerprints on the box. It is just funny and unique quirk of this particular object. =) Speaking about the tree. I personally rather accept not to show visited status for hero at all, that visited it under the command of the other player, than show the precise information for an object, player never visited before. I think this is more fair. But I expect players would be slightly annoyed by that fact, so the current implementation is OK for me. Just my personal opinion. 🙂

The magic garden visually indicates to the player that it has already been visited; the right-click information window indicates the opposite, which makes no sense and is not coherent.

> I'm not saying we should do this, but it seems that the most logical solution to @LeHerosInconnu's issue is to stop showing that a newly recruited hero has visited a tree under the command of another player, thus forcing the player to visit the tree with that new hero to find this out for themself. However, I believe this would be contrary to how all later heroes games work.

This would be contrary to the way the original game works.

In original game:
The hero Ector visited the witch doctor's hut, the gazebo and the tree of knowledge while under the orders of the green player.

https://github.com/user-attachments/assets/04929ee3-551b-43f3-93a1-ea20c895e3b4

The hero Ector is now under the command of the blue player, and it's clearly indicated that these three locations have been visited by Ector.

https://github.com/user-attachments/assets/18d44a59-f61c-456b-99ad-cfe8fc4a29be

The most logical solution is to indicate the additional information added by fheroes2.

Here are the scenario test and save files:
[0 9897b.zip](https://github.com/user-attachments/files/18376810/0.9897b.zip)

**Comment by Branikolog** [comment]
@LeHerosInconnu 

>The magic garden visually indicates to the player that it has already been visited; the right-click information window indicates the opposite, which makes no sense and is not coherent.

From your point - it sounds reasonable. But from the point of overall game logic (If player had visited an object - then show status "visited") everything works fine, since you haven't visited this object actually on this week. The "visited" status could also refer to you exactly (or your particular hero), not to someone, who visited it before you. :)

I suggest leaving everything as it is. Otherwise we have to rework the rest interactive game objects, like graveyards, pyramids, dwellings etc... And this can provide a player an advantage, that was never the case in the OG.

**Comment by LeHerosInconnu** [comment]
@Branikolog,

> @LeHerosInconnu
> 
> > The magic garden visually indicates to the player that it has already been visited; the right-click information window indicates the opposite, which makes no sense and is not coherent.
> 
> From your point - it sounds reasonable. But from the point of overall game logic (If player had visited an object - then show status "visited") everything works fine, since you haven't visited this object actually on this week. The "visited" status could also refer to you exactly (or your particular hero), not to someone, who visited it before you. :)

In this case, the opposing hero recruited has already visited the magic garden during the current week, while under the orders of the opposing player.
This should be updated when the opposing hero is recruited.

> I suggest leaving everything as it is. Otherwise we have to rework the rest interactive game objects, like graveyards, pyramids, dwellings etc... And this can provide a player an advantage, that was never the case in the OG.

The way it's done now is that it's done halfway, because the additional information added by fheroes2 for locations isn't displayed.
Yes, all other locations should also display the additional information known by a recruited opposing hero.
Related discussion: https://github.com/ihhub/fheroes2/discussions/9081.

**Comment by zenseii** [comment]
As I said earlier, this change is very divisive, as has also been shown by the discussion above. I'm therefore going ahead and converting this to a discussion.

**Comment by LeHerosInconnu** [comment]
This is planned in my mod: https://github.com/ihhub/fheroes2/discussions/10103.
I close the discussion.
